clang/lib/CodeGen/CodeGenModule.cpp

```diff
@@ -2059,20 +2057,22 @@ void CodeGenModule::SetLLVMFunctionAttributesForDefinition(const Decl *D,
     B.addAttribute(llvm::Attribute::NoInline);
   }
 
-  // Note: `yk_outline` wins over `yk_unroll_safe`.
-  if (D->hasAttr<YkUnrollSafeAttr>() && !D->hasAttr<YkOutlineAttr>()) {
-    B.addAttribute("yk_unroll_safe");
-    B.addAttribute(llvm::Attribute::NoInline);
-    // FIMXE: check for conflicts with other attr
-  }
-
-  // If the function contains at least one loop, then add an attribute for the
-  // Yk JIT to use for runtime inlining decisions.
-  if (YkHasLoopAnnotations && !F->empty()) {
+  // Mark all functions containing loops with `yk_outline` unless the function
+  // was annotated `yk_unroll_safe`.
+  //
+  // If `F->empty()` then this is merely a function declaration for which we
+  // have no IR. In this case the JIT will be unable to inline calls to this
+  // function anyway, so there's no need to conservatively add `yk_outline`.
+  if (!F->empty()) {
     llvm::DominatorTree DT(*F);
     llvm::LoopInfo LI(DT);
-    if (!LI.empty())
-      B.addAttribute("__yk_has_loop");
+    if (!LI.empty()) {
+      // Note: `yk_outline` wins over `yk_unroll_safe`.
+      if (D->hasAttr<YkUnrollSafeAttr>() && !D->hasAttr<YkOutlineAttr>())
+        B.addAttribute(llvm::Attribute::NoInline);
```

Comment:
Silly question maybe, but this looks like we are adding `NoInline` to every function that isn't `yk_outline`? Shouldn't this be something like:

```
if (D->hasAttr<YkOutlineAttr>() || !D->hasAttr<YkUnrollSafeAttr>) {
  B.addAttribute(llvm::Attribute::NoInline);
  B.addAttribute("yk_outline");
}
```

Reply:
Remind me why we are setting `llvm::Attribute::NoInline` here in the first place. I was just expecting something like:
```
if (!D->hasAttr<YkUnrollSafeAttr>()) {
    B.addAttribute("yk_outline");
}
```

Why do we expect some functions to already have the `YkOutlineAttr` when we get here? I thought this can't be set by the user.

Reply:
> Did you mean to flip && to || in your above code?

Are you referring to me? If so "yes".

But Lukas is right, the `D->hasAttr<YkOutlineAttr>()` is pointless (we'll already have marked that `NoInline` elsewhere),. so we only need `if (!D->hasAttr<YkUnrollSafeAttr>()) { ... }`.

I'm in two minds over `NoInline`. If we *don't* mark functions-with-loops in as `NoInline`, they'll sometimes be inlined and "pollute" calling functions. For us today that will be a performance loss, but in the long run, if the `-O0` version of the interpreter is only used for tracing and deoptimisation, that performance loss will be smaller. On balance, I'm perhaps slightly inclined *not* to add `NoInline` but to mark it as "LOOKAGAINLATER" or something.

Reply:
I understand. And yes, for now we can have `yk_unroll_safe` **not** imply `noinline`.

A question that's taxing me.

Does it matter if a `yk_unroll_safe` function `G` gets inlined into a "not unroll safe" parent function `F` during AOT compilation?

On the one hand `G` will still have its loops unrolled into the parent `F`.

On the other hand, the loops in `G` can't be specialised for the trace, as the trace will be forced to call `F` as an opaque function.

Could just be the ramblings of a madman... :)

Reply:
That's what I meant by "pollute calling functions".

I'm still happy with:

> On balance, I'm perhaps slightly inclined not to add NoInline but to mark it as "LOOKAGAINLATER" or something.

Reply:
Yep, just writing the comment in the code now :)

Reply:
There was actually a case of this "pollution" in one of the tests which slightly changed the shape of an inlined call. I've manually added `noinline` (like in the other test that you commented on) to ensure we check the correct inliner.

Reply:
[932049e](https://github.com/ykjit/ykllvm/pull/44/commits/932049e4914cff5ebea89d3b5c2abc668f3258a7) implements this change, and the follow up commit [973307c](https://github.com/ykjit/ykllvm/pull/44/commits/973307cd8295228e8b1401016858ef99aa734fea) removes some duplication.

Reply:
https://github.com/ykjit/ykllvm/pull/44/commits/932049e4914cff5ebea89d3b5c2abc668f3258a7 looks wrong to me: it's adding `NoInline` even though the comment says it isn't?

Reply:
Note the `!` on the `if` condition.

Reply:
I did note that. My comment stands.

Reply:
I believe it should just be:

```
if (!D->hasAttr<YkUnrollSafeAttr>()) {
  B.addAttribute("yk_outline");
}
```

Reply:
My (initial) worry is that if all calls to the function are inlined at AOT time, then the compiler may discard the function altogether, leaving nothing for the trace to call (I've never seen this happen in practice, but looks like we can get the best of both worlds by annotating it with [used](https://clang.llvm.org/docs/AttributeReference.html#used) instead of `noinline`).

But that aside, I've just commented out the `B.addAttribute(llvm::Attribute::NoInline);` **after the refactoring commit** (which merges the two places where the attributes are added) and then have several failures:
 - `yk_unroll_safe_vs_yk_outline.c`: `yk_outline` no longer takes precedence over `yk_unroll_safe` in the event that the function is annotated with both (you will say, why not emit an error? It's not the clang way, for example you can mark a function `inline` and `alwaysinline` and you will get no error. One attribute silently wins).
 - `no_trace_annotation2.c` and `no_trace_annotation.c` function wasn't outlined as expected. Probably inlined during AOT compilation.

It turns out you that (for now) can skip the `noinline` annotation when handling `yk_unroll_safe`, but not when handling `yk_outline`. This makes my refactoring commit nonsensical.

So my last push reverts the refactoring and removes the `noinline` attribute when handling `yk_unroll_safe`. (I also fixed a bogus comment).

All tests pass.

Reply:
> My (initial) worry is that if all calls to the function are inlined at AOT time, then the compiler may discard the function altogether, leaving nothing for the trace to call (I've never seen this happen in practice, but looks like we can get the best of both worlds by annotating it with [used](https://clang.llvm.org/docs/AttributeReference.html#used) instead of noinline).

Yes, this *definitely* could happen (it's my "polluting" comment above).

*But* https://github.com/ykjit/ykllvm/pull/44/commits/2b36b29e453399ae5009d16556ef09f5003f5546 makes me realise that `yk_outline` also prohivits AOT inlining. So now `yk_outline` stops AOT inlining but `yk_unroll_safe` doesn't. Ug. On reflection (on reflection on reflection!) I have changed my mind again: I think we should simplify our long-term lives (at the expense of short-term performance): `yk_unroll_safe` should also stop AOT inlining.

Sorry for mucking you around!